src/common/Language.ts

```diff
@@ -44,6 +44,11 @@ export type TokenTree = string | TokenTreeArray;
 export type Detokenizer = (tokens: TokenTree) => string;
 export type WhitespaceInsertLogic = (a: string, b: string) => boolean;
 
+export function finiteFlatten(tokenTree: TokenTree): string[] {
+  if (typeof tokenTree === "string") return [tokenTree];
+  return tokenTree.map(finiteFlatten).flat(1);
+}
+
```

Comment:
Why is the function called `finiteFlatten`?

This should be implemented in a way where tokens are being directly appended to the final result array without creating the intermediate arrays.

Reply:
I named it as opposed to the default flat, which TS seemed concerned would be infinite, renamed to `flattenTree` though. Intermediate array creation fixed.